torch/nn/parallel/distributed.py

```diff
@@ -117,14 +122,34 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
         else:
             self._module_copies = [self.module]
 
+        # Split parameters into type buckets so that parameter sync (broadcast)
+        # can operates on mixed parameter types. (e.g. mixed half and float)
+        self.param_type_buckets = {}
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                tp = type(p.data)
+                if tp == torch.cuda.HalfTensor and \
+                        dist._backend != dist.dist_backend.NCCL:
+                    raise RuntimeError("DistributedDataParallel currently only "
+                                       "supports half precision parameters "
+                                       "with NCCL backend")
+                if tp not in self.param_type_buckets:
+                    self.param_type_buckets[tp] = \
+                        [[] for _ in range(len(self.device_ids))]
+                # Add the parameter into the type bucket
+                self.param_type_buckets[tp][device_idx].append(p)
+
         # Split parameters into buckets that will coalesce reductions
-        # TODO: different types need different buckets
-        t = None
-        for p in self.module.parameters():
-            tp = type(p.data)
-            if t is not None and t is not tp:
-                raise ValueError("DistributedDataParallel requires all parameters' data to be of the same type")
-            t = tp
+        #
+        # Note that the NCCL backend currently only supports a single reduction
+        # bucket, so instead of splitting different Tensor types (half, float,
+        # double, etc) into separate buckets, which will form multipel buckets,
+        # we will split the parameters into reduction buckets regardless of
+        # the data types here.
```

Comment:
I don't think this bucketing strategy is good. Consider what will happen if you were to have parameters that are float, then half, then float again, etc. You could easily divide them into separate buckets per type, but now we'll be doing 2x more reductions that we could have

Reply:
This is mainly for the purpose of NCCL backend since it will only supports a single reduction thread. 
Now that NCCL is going to a separate path in https://github.com/pytorch/pytorch/pull/5064/files.  I guess we can directly bucketing the reduction bucket as you said.